package-with/systemd.service

```diff
@@ -7,6 +7,7 @@ After=network.target
 Type=simple
 ExecStart=/usr/sbin/webdavd /etc/webdavd
 LogsDirectory=webdavd
+ProtectSystem=full
```

Comment:
I'm not sure this is necessarily desirable. Certainly `ProtectHome` is undesirable. The primary purpose of this server is to allow access to Linux user's home, so that would be counter productive.  Added protection can be added via the chroot option to protect the system.

However from memory, `ProtectSystem` causes the server to be run in it's own mount namespace. This would be tricky for some users who are unfamiliar with the effect because the server would not give access to (for example) USB drives manually mounted by the user.

Reply:
It would be nice if webdavd could be prevent from home dir access, but then webdav-worker, after authentication as the user, did have access.  But I didn't see an easy way to do that.  ProtectHome prevents authentication from even working.

I don't believe `ProtectSystem` creates a private mount namespace.  There is `PrivateMounts`, but the mount namespace is still propagated in the direction system -> unit private, so usb drives (and the root fs!) still appear.  But mount changes the unit makes will not propagate back to the system.

`ProtectSystem=full` makes some directories read-only, .e.g. /etc, /boot.  Directories were it would be highly unusual to edit files via webdav and most if not all files will be own by root.

All the directories so modified are not reachable after being chrooted to ~, which the default config files does.  So if someone does want to use webdav to get remote root write access to /etc, they already need to edit the webdav config file.  It seems reasonable that they could also edit the unit file too.

Reply:
I've previously had mega problems with this feature (rooted from [this bug](https://github.com/systemd/systemd/issues/7761)) so I'm pretty cautious about it.  Remember this config file is used by everyone who's installed this server for every use case.  It might be worth checking how samba is packaged and if they use this feature.

Regarding the home directory thing, it's certainly not going to fit the general use case.  I don't understand what you're trying to achieve.  I imagine the easiest way to achieve it is to have a single directory somewhere (`/share` or `/user/lib/share`) with everyone's webdavd shares inside and then set the server to chroot.

Reply:
Network servers that I have installed and came with ProtectSystem on:  openvpn-server, systemd-timedated, lldpd, rdisc, chronyd

There are more services that use it, but I've only looked at those that are network daemons.

It's also entirely possible for someone to change the config via a drop in file without even editing this unit file, which is easier than editing the webdavd config file.

Current best practices is to make the default secure.  If someone wants to make it less secure, such as allowing editing their entire file system over webdav, then they can do that.  Making the default insecure and expecting people to secure it themselves isn't recommended anymore.

Example, webdavd isn't even going to work until the default firewall config is edited to allow connections from outside of localhost to port 80 and/or 443.

WRT home dirs, my goal is to be more secure.  There's no reason for webdavd, before authentication of the user, to be able to access any files in home directories.  So it shouldn't have that.  Unfortunately, dropping home dir access can not be undone via authentication as a user.

Reply:
I've also tested the additional mount, by serving `/run`, and then while a webdav session was open, mounting as a user a USB drive.  The new mount appeared to webdavd just fine under /run/username.

This makes /usr read-only too, which doesn't cause a problem starting webdav-worker from /usr/lib or serving the static content in /usr/share.

It does provide an important security benefit knows as W^X.  One can either write files or execute them, not both.  If one is able to get write access to /usr/lib, then one gets the trivial ability to execute arbitrary code, by replacing webdav-worker with a new program which will be conveniently run on the next session.